owlkettle/playground.nim

```diff
@@ -123,165 +126,142 @@ method view (state: DateDialogState): Widget =
     ActionRow:
       title = fieldName
-      Label():
-        text = fmt"Override `toListFormField` for '{$typ.type}'"
-        tooltip = fmt"""
-          The type '{$typ.type}' must implement a `toListFormField` proc:
-          `toListFormField(
-            state: auto, 
-            fieldName: static string, 
-            index: int,
-            typ: typedesc[{$typ.type}]
-          ): Widget`
-          state: The <Widget>State
-          fieldName: The name of the field on `state` that contains the seq for which the current form-row is being generated
-          index: The index on the list of entries in `state.<fieldName>`
-          typ: The type for which `toListFormField` shall function.
-          
-          Implementing the proc will override this dummy Widget.
-          See the playground module for examples.
-        """
-
-proc toListFormField(state: auto, fieldName: static string, index: int, typ: typedesc[ScaleMark]): Widget =
-  ## Provides a form to display a single entry of type `ScaleMark` in a list of `ScaleMark` entries.
-  let mark: ScaleMark = state.getField(fieldName)[index]
-  return gui:
-    ActionRow:
-      title = fieldName & $index
       
       FormulaEntry {.addSuffix.}:
-        value = mark.value
+        value = field[].value
         xAlign = 1.0
         maxWidth = 8
         proc changed(value: float) =
-          state.getField(fieldName)[index].value = value
+          field[].value = value
       
       DropDown {.addSuffix.}:
         items = ScalePosition.items.toSeq().mapIt($it)
-        selected = mark.position.int
+        selected = field[].position.int
         proc select(enumIndex: int) =
-          state.getField(fieldName)[index].position = enumIndex.ScalePosition
-          
-      Button {.addSuffix.}:
-        icon = "user-trash-symbolic"
-        proc clicked() =
-          state.getField(fieldName).delete(index)
+          field[].position = enumIndex.ScalePosition
 
-proc toFormField[T](state: auto, fieldName: static string, typ: typedesc[seq[T]]): Widget =
-  ## Provides a form field for any field on `state` with a seq type.
-  ## Displays a dummy widget if there is no `toListFormField` implementation for type T.
+proc toFormField[T](state: Viewable, field: ptr seq[T], fieldName: string): Widget =
+  ## Provides a form field for any seq type
+  let formFields = collect(newSeq):
+    for index, value in field[]:
+      toFormField(state, field[][index].addr, fmt"{fieldName} {index}")
+      
   return gui:
     ExpanderRow:
       title = fieldName
       
-      for index, num in state.getField(fieldName):
-        insert(toListFormField(state, fieldName, index, T)){.addRow.}
+      for index, formField in formFields:
+        Box(orient = OrientX) {.addRow.}:
+          insert(formField)
+          Button() {.expand: false.}:
+            icon = "user-trash-symbolic"
+            style = [ButtonDestructive]
+            proc clicked() =
+              field[].delete(index) 
```

Comment:
Deleting does not work. GTK outputs `Tried to remove non-child`.

Reply:
That's... new. I literally never encountered this before. However it must've been there for quite a bit because I can't find a commit *now* where that wasn't an issue.

Something super hacky appears in how adding/removing values from `field[]` triggers Widget removals in expander-rows.

Reply:
I did a deeper dive with scale.nim and ScaleMark and *none* of this makes sense.
When clicking the delete button, it correctly changes the value in `field[]`.
That correctly triggers the a re-render, which creates a new set of Widgets that get added to `ExpanderRow`, which then update ExpanderRow's `WidgetState` to replace the old GtkWidgets with the new GtkWidgets in `updateAlignedChildren`.

The error occurs when it tries to call the `remove` proc for the Widget created by `toFormField`. Why that suddenly is the case I can not say, none of this makes sense.
My current suggestion is removing the delete button and forcing users to recompile examples in order to remove list-entries.

Or removing auto-form generation again entirely.

Reply:
To document this in a PR comment at the very least:

The problem was that I was using **Box** to add to `ExpanderRow`, not `ActionRow`.

Can.l pointed out this may be related to https://gitlab.gnome.org/GNOME/libadwaita/-/blob/main/src/adw-preferences-group.c#L381

That is because GTK adds a widget with a different proc in adw_preferences_group_add if it's a child of `PreferencesRow` than it does for anything else. Thus you likely also need to remove it in a different manner, which we currently don't do.